zipkin-server/src/main/java/zipkin2/server/internal/prometheus/ZipkinPrometheusMetricsConfiguration.java

```diff
@@ -129,23 +128,13 @@ public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exc
     AttributeKey.valueOf(Boolean.class, "PROMETHEUS_METRICS_SET");
 
   public static void setup(RequestContext ctx, MeterRegistry registry, String metricName) {
-    if (ctx.hasAttr(PROMETHEUS_METRICS_SET)) {
+    if (ctx.attr(PROMETHEUS_METRICS_SET) == null) {
       return;
     }
-    ctx.attr(PROMETHEUS_METRICS_SET).set(true);
+    ctx.setAttr(PROMETHEUS_METRICS_SET, true);
 
-    ctx.log().addListener(log -> onRequest(log, registry, metricName),
-      RequestLogAvailability.REQUEST_HEADERS,
-      RequestLogAvailability.REQUEST_CONTENT);
-  }
-
-  private static void onRequest(RequestLog log, MeterRegistry registry, String metricName) {
-    Clock clock = registry.config().clock();
-    long startTime = clock.monotonicTime();
-    log.addListener(requestLog -> {
-      getTimeBuilder(requestLog, metricName).register(registry)
-        .record(clock.monotonicTime() - startTime, TimeUnit.NANOSECONDS);
-    }, RequestLogAvailability.COMPLETE);
+    ctx.log().whenComplete().thenAccept(log -> getTimeBuilder(log, metricName).register(registry)
+      .record(log.totalDurationNanos(), TimeUnit.NANOSECONDS));
```

Comment:
I might be missing some subtlety that required using the old version of the code with its own timings instead of Armeria's, let me know.

Reply:
main thing is to keep our Prometheus stable so folks using grafana dont wake up to have their graphs broken. I will look carefully though

Reply:
what you've done is fine. As long as the metric name/dimension matches, it is better to reuse armeria's duration because we don't do much besides invoke a controller anyway.